pkg/util/util.go

```diff
@@ -0,0 +1,346 @@
+	kubeclientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/workqueue"
+
+	types "github.com/kubernetes-sigs/container-object-storage-interface-api/apis/objectstorage.k8s.io/v1alpha1"
+	bucketclientset "github.com/kubernetes-sigs/container-object-storage-interface-api/clientset"
+	"github.com/kubernetes-sigs/container-object-storage-interface-api/controller"
+
+	"sigs.k8s.io/controller-tools/pkg/crd"
+	crdmarkers "sigs.k8s.io/controller-tools/pkg/crd/markers"
+	"sigs.k8s.io/controller-tools/pkg/genall"
+	"sigs.k8s.io/controller-tools/pkg/loader"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+
+	"github.com/golang/glog"
+)
+
+var (
+	// Error codes that the central controller will return
+	ErrStopProvision  = errors.New("stop provisioning")
+	ErrBCUnavailable  = errors.New("BucketClass is not available")
+	ErrNotImplemented = errors.New("Operation Not Implemented")
+)
+
+func CopySS(m map[string]string) map[string]string {
+	if m == nil {
+		return nil
+	}
+	copy := make(map[string]string, len(m))
+	for k, v := range m {
+		copy[k] = v
+	}
+	return copy
+}
+
+func CopyStrings(s []string) []string {
+	if s == nil {
+		return nil
+	}
+	c := make([]string, len(s))
+	copy(c, s)
+	return c
+}
+
+func GetUUID() string {
+	return string(uuid.NewUUID())
+}
+
+func SetupTest(ctx context.Context) (bucketclientset.Interface, kubeclientset.Interface, *controller.ObjectStorageController) {
+
+	// Initialize the clients
+	config, err := func() (*rest.Config, error) {
+		kubeConfig := os.Getenv("KUBECONFIG")
+		if kubeConfig != "" {
+			return clientcmd.BuildConfigFromFlags("", kubeConfig)
+		}
+		return rest.InClusterConfig()
+	}()
+	if err != nil {
+		glog.Fatalf("Failed to create clients: %v", err)
+	}
+
+	kubeClient, err := kubeclientset.NewForConfig(config)
+	if err != nil {
+		glog.Fatalf("Failed to create Kubernetes client: %v", err)
+	}
+	client, err := bucketclientset.NewForConfig(config)
+	if err != nil {
+		glog.Fatalf("Failed to create ObjectStorage client: %v", err)
+	}
+
+	crdClientset, err := apiextensions.NewForConfig(config)
+	if err != nil {
+		glog.Fatalf("Failed to create CRD client: %v", err)
+	}
+
+	err = RegisterCRDs(ctx, crdClientset.CustomResourceDefinitions())
+	if err != nil {
+		glog.Fatalf("Failed to register CRDs: %v", err)
+	}
+
+	rateLimit := workqueue.NewMaxOfRateLimiter(
+		workqueue.NewItemExponentialFailureRateLimiter(100*time.Millisecond, 600*time.Second),
+		&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
+	)
+	ctrl, err := controller.NewObjectStorageControllerWithClientset("controller-manager", "leader-lock", 40, rateLimit, kubeClient, client)
+	if err != nil {
+		glog.Fatalf("Failed to create ObjectStorage Controller: %v", err)
+	}
+
+	go ctrl.Run(ctx)
+	return client, kubeClient, ctrl
+}
+
+func GetBuckets(ctx context.Context, client bucketclientset.Interface, numExpected int) *types.BucketList {
```

Comment:
This is likely only used in tests? Could use a comment at least to explain what the function does, how it behaves.